serix/serix.go

```diff
@@ -0,0 +1,728 @@
+// During the encoding process serix also performs the validation if such option was provided.
+// Use the options list opts to customize the serialization behaviour.
+// To ensure deterministic serialization serix automatically applies lexical ordering for maps.
+func (api *API) Encode(ctx context.Context, obj interface{}, opts ...Option) ([]byte, error) {
+	value := reflect.ValueOf(obj)
+	if !value.IsValid() {
+		return nil, errors.New("invalid value for destination")
+	}
+	opt := &options{}
+	for _, o := range opts {
+		o(opt)
+	}
+	return api.encode(ctx, value, opt.ts, opt)
+}
+
+// Decode deserializer bytes b into the provided object obj.
+// obj must be a non-nil pointer for serix to deserialize into it.
+// serix traverses the object recursively and deserializes everything based on its type.
+// If a type implements the custom Deserializable interface serix delegates the deserialization the that type.
+// During the decoding process serix also performs the validation if such option was provided.
+// Use the options list opts to customize the deserialization behavior.
+func (api *API) Decode(ctx context.Context, b []byte, obj interface{}, opts ...Option) (int, error) {
+	value := reflect.ValueOf(obj)
+	if !value.IsValid() {
+		return 0, errors.New("invalid value for destination")
+	}
+	if value.Kind() != reflect.Ptr {
+		return 0, errors.Errorf(
+			"can't decode, the destination object must be a pointer, got: %T(%s)", obj, value.Kind(),
+		)
+	}
+	if value.IsNil() {
+		return 0, errors.Errorf("can't decode, the destination object %T must be a non-nil pointer", obj)
+	}
+	value = value.Elem()
+	opt := &options{}
+	for _, o := range opts {
+		o(opt)
+	}
+	return api.decode(ctx, b, value, opt.ts, opt)
+}
+
+// RegisterValidators registers validator functions that serix will call during the Encode and Decode processes.
+// There are two types of validator functions:
+//
+// 1. Syntactic validators, they validate the Go object and its data.
+// For Encode they are called for the original Go object before serix serializes the object into bytes.
+// For Decode they are called after serix builds the Go object from bytes.
+//
+// 2. Bytes validators, they validate the corresponding bytes representation of an object.
+// For Encode they are called after serix serializes Go object into bytes
+// For Decode they are called for the bytes before serix deserializes them into a Go object.
+//
+// The validation is called for every registered type during the recursive traversal.
+// It's an early stop process, if some validator returns an error serix stops the Encode/Decode and pops up the error.
+//
+// obj is an instance of the type you want to provide the validator for.
+// Note that it's better to pass the obj as a value, not as a pointer
+// because that way serix would be able to derefer pointers during Encode/Decode
+// and detect the validators for both pointers and values
```

Comment:
It will be error-prone if you register your validators/type settings for values. 
It's a common practice.
Otherwise, if you register it for a pointer, and during Encode() serix encounters a value type, it can't get a pointer from it.

Reply:
I see what you mean. To always derefer at the registration stage. 
Yes, we could do that, but we client also needs to pass a validator func with the first argument of the same type as the object it's been registered for. 
If we deref the object in the registration when it wouldn't match with the func first argument. 
It complicates stuff, I think it's not worth it

Reply:
Ok then we leave it like it is